src/storage/sql_table.cpp

```diff
@@ -19,64 +43,329 @@ SqlTable::SqlTable(BlockStore *const store, const catalog::Schema &schema, const
+  for (auto &it : map) {
+    redo_col_oids.emplace_back(it.first);
+    // check if the col_oid exists in the old schema
+    if (tables_.Find(old_version)->second.column_map.count(it.first) == 0) {
+      is_subset = false;
+      break;
+    }
+  }
+
+  storage::TupleSlot ret_slot;
+  if (is_subset) {
+    // we can update in place
+    auto old_dt_version = tables_.Find(old_version)->second;
+
+    // The slot version is not the same as the version_num
+    col_id_t original_column_ids[redo->NumColumns()];
+    ModifyProjectionHeaderForVersion(redo, tables_.Find(version_num)->second, old_dt_version, original_column_ids);
+
+    // Get the result and copy back the old header
+    bool result = old_dt_version.data_table->Update(txn, slot, *redo);
+
+    // We should create a buffer of old Projected Row and update in place. We can't just
+    // directly erase the data without creating a redo and update the chain.
+
+    // auto old_pair = InitializerForProjectedRow(redo_col_oids, old_version);
+
+    // // 1. Create a ProjectedRow Buffer for the old version
+    // byte *buffer = common::AllocationUtil::AllocateAligned(old_pair.first.ProjectedRowSize());
+    // storage::ProjectedRow *pr = old_pair.first.InitializeRow(buffer);
+
+    // // 2. Copy from new ProjectedRow to old ProjectedRow
+    // StorageUtil::CopyProjectionIntoProjection(redo, map, tables_.Find(old_version)->second.layout, pr,
+    // old_pair.second);
+
+    // // 3. Update the old data-table
+    // bool result = tables_.Find(old_version)->second.data_table->Update(txn, slot, *pr);
+    // delete[] buffer;
+    if (!result) {
+      return {false, slot};
+    }
+    ret_slot = slot;
+  } else {
+    STORAGE_LOG_DEBUG("have to delete and insert ... ");
+
+    // need to create a new ProjectedRow of all columns
+    // 1. Get the old row
+    // 2. Convert it into new row
+    // 3. Delete old row
+    // 4. Update the new row before insert
+    // 5. Insert new row into new table
+
+    // 1. Get old row
+    // 2. Convert it into new row
+    std::vector<catalog::col_oid_t> new_col_oids;  // the set of col oids which the new schema has
+    for (auto &it : tables_.Find(version_num)->second.column_map) new_col_oids.emplace_back(it.first);
+    auto new_pair = InitializerForProjectedRow(new_col_oids, version_num);
+    auto new_buffer = common::AllocationUtil::AllocateAligned(new_pair.first.ProjectedRowSize());
+    ProjectedRow *new_pr = new_pair.first.InitializeRow(new_buffer);
+    bool valid = Select(txn, slot, new_pr, new_pair.second, version_num);
+    if (!valid) {
+      delete[] new_buffer;
+      return {false, slot};
+    }
+    // 3. Delete the old row
+    bool succ = tables_.Find(old_version)->second.data_table->Delete(txn, slot);
+
+    // 4. Update the new row before insert
+    StorageUtil::CopyProjectionIntoProjection(*redo, map, tables_.Find(version_num)->second.layout, new_pr,
```

Comment:
Do you need to populate default values for Update too?
Also in CopyProjectionIntoProjection you called here, there is a TODO listing for default values, and I think you can remove it and add something here outside.

Reply:
The full materialization of the existing data between line 256 and line 265 will handle default values because we use the SqlTable `Select` to materialize it into the version we will insert.